### PR TITLE
9333 default observer fix

### PR DESCRIPTION
**Pull request by jlitzingerdev**
This is a fix for 9333: https://twistedmatrix.com/trac/ticket/9333

Some items of note for reviewers:

 1.  The new formatEventWithTracebacks API was deliberately kept private to avoid adding new public APIs to maintain. Additionally, its easy to make it public if so specified in review.

2.  The existing formatEventAsClassicLogText has similar code that can be replaced. Rather than assume that was desired, I kept the change as focused as possible. Easy to change if desired.

As noted in the ticket, there are cases where duplicate tracebacks result, specifically if observers are appended to the globalLogPublisher, rather than overwrite with a beginLoggingTo call on the globalLogBeginner. This is documented in the logger docs.

### CODE REVIEW

**Comment by hawkowl** [comment]
Pushed to the buildbots.

**Comment by hawkowl** [comment]
The OSX build failures are spurious. They pass in circlecli.

File: src/twisted/logger/_format.py
Comment (markrwilliams) [inline src/twisted/logger/_format.py]: This should be lower case and contain the exception message, like [its counterpart in the legacy logging system](https://github.com/twisted/twisted/blob/104394ddb76775c57c18897160a5ad055617b51a/src/twisted/python/log.py#L447).

File: src/twisted/logger/_format.py
Comment (markrwilliams) [inline src/twisted/logger/_format.py]: This should be `except Exception:` - we don't want to catch `SystemExit`!

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This is an incomplete sentence.  I think it should be, "... all of which ensure that the returned type is UTF-8 decoded text."

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: A better test would be:

```python
try:
    raise CapturedError("This is a fake error")
except CapturedError:
    f = Failure()
...
self.assertIn(f.getTraceback().decode('utf-8'), eventText)
...
```

That's because it would exercise the code with a real traceback.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: As above.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This is extraneous to the stated purpose of the test, which is served by `{evil()}`.  Please remove it.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This should be a unicode literal.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This should be a unicode literal.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This should be a unicode literal.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This should be a unicode literal.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This should be a unicode literal.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This should be a unicode literal.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This should be a unicode literal - also see above comment about the message in this case.

File: src/twisted/logger/test/test_format.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_format.py]: This should be a unicode literal.

File: src/twisted/logger/test/test_global.py
Comment (markrwilliams) [inline src/twisted/logger/test/test_global.py:370]: This (and what follows) should _not_ be a unicode literal :)

That's because [FileLogObserver encodes unicode messages to UTF-8 when it's given a binary file](https://github.com/twisted/twisted/blob/104394ddb76775c57c18897160a5ad055617b51a/src/twisted/logger/_file.py#L33-L37), and [it is on Python 2](https://github.com/jlitzingerdev/twisted/blob/a0a1c7c2bd7b7a2f396ea6ddb0453c131ccec4fc/src/twisted/logger/test/test_global.py#L65).

**Review by markrwilliams — APPROVED** [review]
This is an important improvement and a well-done PR.  Thank you so much!

Please address the review comments before resubmitting for review.

**Review by markrwilliams — CHANGES_REQUESTED** [review]
This review is to block merging until comments are addressed.

File: src/twisted/logger/_format.py
Comment (jlitzingerdev) [reply src/twisted/logger/_format.py]: No problem.  I modeled formatEventWithTraceback after formatEventAsClassicLogText (above it), which also uses all caps and an bare except.  Is that something that should be fixed (in a different ticket/PR)?

Comment (markrwilliams) [reply src/twisted/logger/_format.py]: Yes, I think `formatEventAsClassicLogText` should get changed via a new ticket.  I appreciate that you started with a goal of consistency, as well!

File: src/twisted/logger/test/test_format.py
Comment (jlitzingerdev) [reply src/twisted/logger/test/test_format.py]: Makes sense.  What I submit will be slightly different as eventText is always unicode and getTraceback() is the native string type (the decode call fails on P3).

File: src/twisted/logger/test/test_format.py
Comment (jlitzingerdev) [reply src/twisted/logger/test/test_format.py]: @markrwilliams I agree, with concerns.  It tests a valid "unformattable" path, but also neglects one.  The first path would be caught by your suggested change, eliminating the `Unformattable()`.  The second path, exercised by the `Unformattable()` key occurs when the __repr__ call for the event throws.  My question -- do we only need to exercise the former, or should I split this into two tests, one for each?  For the sake of completeness, that seems like the right thing to do, thoughts?

File: src/twisted/logger/_format.py
Comment (jlitzingerdev) [reply src/twisted/logger/_format.py]: @markrwilliams As with the bare except, I modeled this after the existing similar example in `formatEventAsClassicLogText`, rather than the legacy log system.  Should `formatEventAsClassicLogText` be changed to match the legacy (and concatenate the error) in a separate ticket?

File: src/twisted/logger/test/test_global.py
Comment (jlitzingerdev) [reply src/twisted/logger/test/test_global.py:370]: @markrwilliams Am I correct in interpreting that this code should remain as-is, to exercise the different paths for a native string on Python2/Python3?  Or, did you want me to ensure the encoding was forced on both?

Comment (markrwilliams) [reply src/twisted/logger/test/test_global.py:370]: Sorry for being unclear!  This was more of a note to myself confirming that this code is correct to use a native string.

File: src/twisted/logger/_format.py
Comment (jlitzingerdev) [inline src/twisted/logger/_format.py:311]: I have concerns with this code.  Failure's getTraceback() method will return a native string depending on the Python version, so str/bytes Python 2.  The assumption that the bytes can sanely decoded as utf-8 is not one I've been able to verify and is probably wrong.

Looking at `formatEventAsClassicLogText`, which I used as a guide, it doesn't do any conversion but seems (I may be misreading) to rely on the fact that bytes and unicode can be joined in Python2 (https://github.com/twisted/twisted/blob/2e7b403c03e9b9b891a5eb36297e494f541d5d6a/src/twisted/logger/_format.py#L192).  That may (haven't confirmed) expose it to the same issue as 9442 due to the ascii conversion.  Should I confirm that I'll open a ticket.

So while this might be fine, I'd like to do a bit more diligence before this is merged, thus, I'm removing it from review.

Comment (jlitzingerdev) [reply src/twisted/logger/_format.py:311]: Related to the concern with `formatEventAsClassicLogText`:
https://twistedmatrix.com/trac/ticket/9456#ticket

**Comment by jlitzingerdev** [comment]
Closing, no longer relevant.

**Pull request closed without merging**
